### PR TITLE
first pass at typing

### PR DESCRIPTION
**Pull request by adamlerer**
some initial attempts at adding type hints. Some bigger refactors will be required before these start showing their usefulness, but this gives the basic idea for feedback.

### CODE REVIEW

**Comment by vineetbansal** [comment]
Thanks @adamlerer - this looks great!

It seems that `pyright` wasn't actually run on this branch though, is that correct? (I'll tag a couple of places where it is failing, for example). I was thinking while merging this, we also add `pyright` as a `[dev]` dependency, add a minimal (possibly empty to being with) section `[tool.pyright]` in our `pyproject.toml`, make it part of the `pre-commit` checks + have it run as part of the CI. This should still allow for incremental annotations, while protecting us against *incorrect* ones.

So perhaps we should fix the `pyright` errors first? What do you think?

File: cryodrgn/dataset.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 log = utils.log
 
 
-def load_particles(mrcs_txt_star, lazy=False, datadir=None):
+def load_particles(mrcs_txt_star: str, lazy: bool = False, datadir: bool = None):
```
Comment (vineetbansal) [inline cryodrgn/dataset.py:19]: One place where `pyright` fails - cannot have default `None` if the type is `bool`.

**Comment by adamlerer** [comment]
Good questions. My experience has been that it's challenging to get a whole repo to pass pyright right off the bat because of type checks in underlying libraries. What worked for me was to add pyright to CI checks with a whitelist or blacklist that we can incrementally improve. Let me know what you think!

I'm still trying to understand whether enforcing types is going to be more of a benefit or a harm in cryoDRGN, since the current codebase uses dynamic / duck typing extensively, and I don't want to slow down the development process by enforcing type consistency. For example, in lots of places its unclear whether the inputs/outputs of a function should be numpy or torch, which on the one hand is *super* annoying and slows down development, on the other hand makes the codebase more flexible.

**Comment by vineetbansal** [comment]
```
it's challenging to get a whole repo to pass pyright right off the bat..
```
I don't think this is true - any type checker (including `pyright`) will always pass in the absence of any annotations, and will always pass as one incrementally starts to add annotations, which, while not completely specified across all callers/callees, are at least mutually consistent where they *are* specified (the confirmation of which was the whole point of doing all this anyway).

I agree about the library code we use - it's not worth checking right off the bat - I'm sure we can ask `pyright` to only check code in our repo but the correct specification of `include/exclude` to start with, if its doing that by default. The whitelist should always include all our code though, otherwise we're just saying - "I know this doesn't pass, I'll get to it later..". If one is going to defer fixing errors as `pyright` reports them, one might as well defer adding annotations at that place at all.

```
whether enforcing types is going to be more of a benefit or a harm in cryoDRGN
```

I think it's always a benefit, never a harm. In the case I pointed out above where it's failing, it forces us to think what we're telling the callers of a function. The `pyright` error above tells me that I should make explicit in the annotation that callers are allowed to pass `None`. The other error around:
```
datadir = os.path.dirname(
```
told me that `datadir` was perhaps meant to be processed as a string, so my annotation is incorrect. Both excellent findings, which would be lost if we started whitelisting/blacklisting the pyright-visible parts of our code.

**Comment by adamlerer** [comment]
Cool, lets do it! 

> I agree about the library code we use - it's not worth checking right off the bat - I'm sure we can ask pyright to only check code in our repo using the correct specification of include/exclude to start with, if its doing that by default.

I actually find that the most benefit comes from when you use the type annotations in the libraries you use as well. But I'm free to turn that on in my extension for now and leave it off in CI.

One thing we should add are README instructions for setting up vscode and possibly other IDEs with the right settings for our codebase (mainly auto-black and pyright type checking).

**Comment by vineetbansal** [comment]
@adamlerer - ok - let me take your branch and weed out the current `pyright` errors today so we can start with a clean slate. I will send a PR towards your `typing` branch soon.

I don't personally use VsCode (but Pycharm) so I can write up some stuff for Pycharm. Perhaps you can tackle Vscode. A couple of weeks ago I added a Wiki page for `Developer Setup` since I thought the README is for a more general audience. Maybe that's a good place for this kind of stuff?

**Comment by vineetbansal** [comment]
Update: I've been at this many hours but there still remains some work to be done. Probably another day. I didn't realize that unlike `mypy`, `pyright` reports a lot of problems right off the bat, _even in the absence of any annotations at all._ That's great I suppose, but adding to that the fact that we're starting off with these annotations, its not trivial to remove all those errors on short notice.

@adamlerer - if the initial batch of types that you put in your branch were well thought-out and you'd like to keep them as a starting point (which I'm assuming is the case), I'll keep going this route. The only thing to keep in mind is that many hundreds/thousands of lines of code will be touched in the process of fixing them, so I hope the PR can be reviewed+merged soon, otherwise it will get tricky to merge later. If you're okay starting from scratch, we can perhaps start incrementally (without any annotations), get the CI and other checks in, make them pass, call that the starting point, and then start introducing annotations slowly.

**Comment by adamlerer** [comment]
@vineetbansal can you post a branch with what you've done so far? I'm curious.

I wouldn't say my choice of things to annotate was "well though out", it was more just the pieces that were easiest to do. A lot of other parts of the codebase require more substantial refactoring to enforce type hygiene.

Re thousands of lines of code:
1. Did you enable strict or basic type checking? I would only recommend basic type checking - strict is way too strict
2. In some places I find it's better to add a `# type: ignore` pragma rather than "fixing" the types
3. You might want to consider excluding some files (for now at least) from type checking.

**Comment by vineetbansal** [comment]
Closing this since `pyright` stuff is now in through a separate PR.

**Pull request closed without merging**